02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,278 @@
+# References and Values
+
+## Learning Goals
+At the end of this lesson, students should be able to...
+
+- Discuss how objects are stored in Ruby
+- Differentiate between _references_ and _values_
+- Compare _modifying an object_ with _reassigning a variable_
+
+## Motivation
+
+We're going to start today with some Ruby code that does something a little unexpected. It's a method that takes an array of strings as an argument, and truncates (chops off the end of) all the strings with more than three characters. Or at least that's the idea.
+
+```ruby
+def short_strings(input)
+  result = []
+  input.each do |word|
+    # Slice characters 0 to 2
+    result << word[0..2]
+  end
+  input = result
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings(pets)
+puts "#{pets}"
+
+```
+
+Running this code results in `["dog", "parrot", "cat", "llama"]`. The array in unchanged! Let's do some debugging:
+
+```ruby
+def short_strings(input)
+  result = []
+  input.each do |word|
+    # Slice characters 0 to 2
+    result << word[0..2]
+  end
+  input = result
+  puts "Inside short_strings, input is"
+  puts "#{input}"
+end
+
+pets = ['dog', 'parrot', 'cat', 'llama']
+short_strings(pets)
+puts "After calling short_strings"
+puts "#{pets}"
+```
+
+The results:
+
+```
+Inside short_strings, input is
+["dog", "par", "cat", "lla"]
+After calling short_strings
+["dog", "parrot", "cat", "llama"]
+```
+
+`short_strings` is indeed creating a list of shortened words and storing it under the name `input`, but our outer variable `pets` isn't being updated. The reason why has to do with references and values, and how data is stored in a computer.
+
+As an aside: one way to fix our method is to return the new array, and when calling it say `pets = short_strings(pets)`. However, sometimes this isn't an option - for example, what if our method was supposed to return the number of words that had to be truncated?
+
+## References and Values
+
+When we create an array in Ruby (or a string or a hash or any other complex data type), we're actually creating two things.
+
+The first is the _value_ of the array, which involves asking the operating system for a bit of memory and then putting our data in it. You can think of this as the actual object. Each piece of memory we get from the OS has an _address_ representing its physical location in hardware, which is how we get back to it later.
```

Comment:
I think we could simplify these two paragraphs by removing the details about memory allocation.

For example, we could say that the _value_ of the array is "the actual data that our computer stores somewhere" and a _reference_ to the array is the "location where the actual data can be found."

Divorcing the discussion of references and values from the actual implementation Ruby uses would also make it easier to provide examples of other places where we use references:

| Reference | Value |
|:-----------|:-------|
| URL | Web page or file |
| File path on hard drive | The contents of that file |
| ID number in CSV | Row of data with that ID in another CSV file |

This would, however, necessitate changing a number of the excellent diagrams you've created.